src/pipelinedefinition.cpp

```diff
@@ -85,14 +87,70 @@ Status PipelineDefinition::validate(ModelManager& manager) {
     return validationResult;
 }
 
+Status PipelineDefinition::initializeNodeResources() {
+    for (const auto& nodeInfo : nodeInfos) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            void* customNodeLibraryInternalManager = nullptr;
+            auto it = nodeResources.find(nodeInfo.nodeName);
+            bool internalManagerExists = (it != nodeResources.end());
+            if (internalManagerExists) {
+                customNodeLibraryInternalManager = it->second;
+            }
+            auto params = createCustomNodeParamArray(nodeInfo.parameters).get();
+            if (!nodeInfo.library.isValid()) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Pipeline: {} node: {} refers to invalid library", pipelineName, nodeInfo.nodeName);
+                return StatusCode::PIPELINE_DEFINITION_INVALID_NODE_LIBRARY;
+            }
+            auto status = nodeInfo.library.initialize(&customNodeLibraryInternalManager, params, nodeInfo.parameters.size());
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Initialization of library with base path: {} failed", nodeInfo.library.basePath);
+                return StatusCode::NODE_LIBRARY_INITIALIZE_FAILED;
+            }
+            if (!internalManagerExists) {
+                nodeResources.insert({nodeInfo.nodeName, customNodeLibraryInternalManager});
+            }
+        }
+    }
+    return StatusCode::OK;
+}
+
+std::vector<NodeInfo> PipelineDefinition::calculateNodeInfosDiff(const std::vector<NodeInfo>& nodeInfos) {
+    std::vector<NodeInfo> diff;
+    for (const auto& nodeInfo : this->nodeInfos) {
+        auto it = std::find_if(nodeInfos.begin(), nodeInfos.end(),
+            [&nodeInfo](const auto& x) { return x.nodeName == nodeInfo.nodeName; });
+        if (it == nodeInfos.end()) {
+            diff.push_back(nodeInfo);
+        }
+    }
+    return diff;
+}
+
+void PipelineDefinition::deinitializeNodeResources(const std::vector<NodeInfo>& nodeInfosDiff) {
+    for (const auto& nodeInfo : nodeInfosDiff) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            auto it = nodeResources.find(nodeInfo.nodeName);
+            if (it == nodeResources.end()) {
+                continue;
```

Comment:
I think we might log internal error here, reaching this place in code indicates error in ovms.